File: kt/2019/day02/day02.kt
package _2019.day02

import _2019.intcode.IntcodeComputer
import java.io.File

fun main() {
    val data = File("./kt/2019/day02/input.txt")
            .readText()
            .split(",")
            .map { it.toInt() }

    println("Solution 1: ${solution1(data)}")
    println("Solution 2: ${solution2(data)}")
}

fun solution1(program: List<Int>, noun: Int = 12, verb: Int = 2): String {
    val computer = IntcodeComputer(program)
    val memory = computer.getMemoryDangerously()
    memory[1] = noun
    memory[2] = verb
    computer.runUntilBlocked()
    return computer.getMemory().getValue(0).toString(10)
}

fun solution2(program: List<Int>): String {
    (1..100).forEach { noun ->
        (1..100).forEach { verb ->
            if (solution1(program, noun, verb) == "19690720") {
                return ((noun * 100) + verb).toString()
            }
        }
    }
    return "Unable to find noun-verb combo!"
}

File: py/2019/day02/day02.py
from typing import List

from intcode import IntcodeComputer


def solution1(data: List[int], noun=12, verb=2) -> int:
    computer = IntcodeComputer(data)
    computer.registers[1] = noun
    computer.registers[2] = verb
    while not computer.halted:
        computer.run()
    return computer.registers[0]


def solution2(data: List[int]) -> int:
    target = 19690720
    for noun in range(100):
        for verb in range(100):
            result = solution1(data, noun=noun, verb=verb)
            if result == target:
                return 100 * noun + verb


if __name__ == "__main__":
    with open("input.txt", "r") as f:
        data = list(map(int, f.readline().split(",")))
    print(f"Solution 1: {solution1(data)}")
    print(f"Solution 2: {solution2(data)}")
